// @stackbit/sdk/src/analyzer/analyze-schema-types.ts
/**
 * This file extends the Stackbit schema with an "unknown" field type.
 * The "unknown" field type is used in intermediate steps when generating a new
 * schema from content files
 */

import {
    FieldType,
    FieldCommonProps,
    FieldBasicProps,
    FieldEnumProps,
    FieldImageProps,
    FieldNumberProps,
    FieldObjectProps,
    FieldModelProps,
    FieldReferenceProps,
    FieldCrossReferenceProps,
    FieldListProps,
    FieldListItems
} from '@stackbit/types';

export type FieldWithUnknown =
    | FieldUnknown
    | FieldSimpleWithUnknown
    | FieldEnumWithUnknown
    | FieldImageWithUnknown
    | FieldNumberWithUnknown
    | FieldObjectWithUnknown
    | FieldModelWithUnknown
    | FieldReferenceWithUnknown
    | FieldCrossReferenceWithUnknown
    | FieldListWithUnknown;

export type FieldUnknown = FieldCommonProps & FieldUnknownProps;
export type FieldSimpleWithUnknown = FieldCommonProps & FieldBasicProps;
export type FieldEnumWithUnknown = FieldCommonProps & FieldEnumProps;
export type FieldImageWithUnknown = FieldCommonProps & FieldImageProps;
export type FieldNumberWithUnknown = FieldCommonProps & FieldNumberProps;
export type FieldObjectWithUnknown = FieldCommonProps & FieldObjectPropsWithUnknown;
export type FieldModelWithUnknown = FieldCommonProps & FieldModelProps;
export type FieldReferenceWithUnknown = FieldCommonProps & FieldReferenceProps;
export type FieldCrossReferenceWithUnknown = FieldCommonProps & FieldCrossReferenceProps;
export type FieldListWithUnknown = FieldCommonProps & FieldListPropsWithUnknown;

export type FieldTypeWithUnknown = 'unknown' | FieldType;

export interface FieldUnknownProps {
    type: 'unknown';
}

export type FieldObjectPropsWithUnknown = Omit<FieldObjectProps, 'fields'> & {
    fields: FieldWithUnknown[];
};

export type FieldListPropsWithUnknown = Omit<FieldListProps, 'items'> & {
    items?: FieldListItemsWithUnknown;
};

export type FieldListItemsWithUnknown = Exclude<FieldListItems, FieldObjectProps> | FieldUnknownProps | FieldObjectPropsWithUnknown;

export function isObjectWithUnknownField(field: FieldWithUnknown): field is FieldObjectWithUnknown {
    return field.type === 'object';
}

export function isListWithUnknownField(field: FieldWithUnknown): field is FieldListWithUnknown {
    return field.type === 'list';
}

export function isObjectListItemsWithUnknown(items: FieldListItemsWithUnknown): items is FieldObjectPropsWithUnknown {
    return items.type === 'object';
}

export function isModelListItemsWithUnknown(items: FieldListItemsWithUnknown): items is FieldModelProps {
    return items.type === 'model';
}
