// @stackbit/sdk/dist/analyzer/site-analyzer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.analyzeSite = void 0;
const lodash_1 = __importDefault(require("lodash"));
const file_browser_1 = require("./file-browser");
const ssg_matcher_1 = require("./ssg-matcher");
const cms_matcher_1 = require("./cms-matcher");
const schema_generator_1 = require("./schema-generator");
async function analyzeSite(options) {
    const fileBrowser = file_browser_1.getFileBrowserFromOptions(options);
    const ssgMatchResult = await ssg_matcher_1.matchSSG({ fileBrowser });
    const cmsMatchResult = await cms_matcher_1.matchCMS({ fileBrowser });
    let schemaGeneratorResult = null;
    if (!cmsMatchResult) {
        schemaGeneratorResult = await schema_generator_1.generateSchema({ ssgMatchResult, fileBrowser });
    }
    const dataDir = (ssgMatchResult === null || ssgMatchResult === void 0 ? void 0 : ssgMatchResult.dataDir) !== undefined ? ssgMatchResult.dataDir : schemaGeneratorResult === null || schemaGeneratorResult === void 0 ? void 0 : schemaGeneratorResult.dataDir;
    const pagesDir = (ssgMatchResult === null || ssgMatchResult === void 0 ? void 0 : ssgMatchResult.pagesDir) !== undefined ? ssgMatchResult.pagesDir : schemaGeneratorResult === null || schemaGeneratorResult === void 0 ? void 0 : schemaGeneratorResult.pagesDir;
    const assets = generateAssets(ssgMatchResult);
    let config = {
        stackbitVersion: '~0.3.0',
        ssgName: ssgMatchResult === null || ssgMatchResult === void 0 ? void 0 : ssgMatchResult.ssgName,
        cmsName: cmsMatchResult === null || cmsMatchResult === void 0 ? void 0 : cmsMatchResult.cmsName,
        nodeVersion: ssgMatchResult === null || ssgMatchResult === void 0 ? void 0 : ssgMatchResult.nodeVersion,
        publishDir: ssgMatchResult === null || ssgMatchResult === void 0 ? void 0 : ssgMatchResult.publishDir,
        dataDir: dataDir,
        pagesDir: pagesDir,
        assets: assets,
        models: (schemaGeneratorResult === null || schemaGeneratorResult === void 0 ? void 0 : schemaGeneratorResult.models) || [],
        dirPath: '/temp',
        filePath: '/temp/stackbit.yaml'
    };
    config = lodash_1.default.omitBy(config, lodash_1.default.isUndefined);
    return {
        ssgMatchResult,
        cmsMatchResult,
        config
    };
}
exports.analyzeSite = analyzeSite;
function generateAssets(ssgMatchResult) {
    if ((ssgMatchResult === null || ssgMatchResult === void 0 ? void 0 : ssgMatchResult.assetsReferenceType) === 'static' && (ssgMatchResult === null || ssgMatchResult === void 0 ? void 0 : ssgMatchResult.staticDir)) {
        return {
            referenceType: 'static',
            staticDir: ssgMatchResult === null || ssgMatchResult === void 0 ? void 0 : ssgMatchResult.staticDir,
            uploadDir: 'assets',
            publicPath: '/'
        };
    }
    return undefined;
}
